### ISSUE TITLE
Hosting third-party domain

### ISSUE THREAD

**Issue by Jehan** [Chocobozzz/PeerTube#3658]
**Describe the problem to be solved**

Similar to how some email hosting providers can propose a specific "custom domain name" service, it would be very interesting for Peertube to have such options (which can be enabled or not of course, and instance would have the ability to put a price on such additional service as it is additional work and administration). Typically for email, say you own example.org but you don't want to administer your own email servers. Some email provides email hosting for your domain name.

Say now you want to have your own Peertube instance with your domain name, but not administer it yourself (or simply you are not a system admin, or are a very small company with no admin, etc.). The advantages are twofold:

1. You can tell people to go to your domain name which would redirect on your channel(s) on the shared hosting. Looks more "professional".
2. Your video links could use your domain so that if you ever want to change provider, you would want to be able to move all your videos to your new hosting, not break any links. You could also end up self-hosting some day and keep the whole video catalogue as well as the existing links. This gives you independence and choosing an initial provider would be less scary because wherever you go, you keep your Peertube handle and your links can stay alive.

**Describe the solution you would like:**

* You could subscribe to an existing Peertube provider, for instance example.com.
* If example.com activated the option and you own example.org, you could ask to join example.com with an example.org login. So your login would actually be nickname@example.org (instead of nickname@example.com).
* Of course, you have to set up your DNS servers to redirect example.org to example.com.
* When requesting the example.org page, example.com would show a custom page as configured by nickname@example.org (possibly another logo, and your video would be displayed by default instead of the whole shared peertube videos, etc.).
* If some day, you want to leave example.com and join another Peertube provider (or host your own), you could just move all your videos automatically with a specific protocol to be defined where example.com would check that the new host really has authorization. All links should stay the same.

**Comment by kontrollanten** [comment]
Interesting idea!

> When requesting the example.org page, example.com would show a custom page as configured by nickname@example.org (possibly another logo, and your video would be displayed by default instead of the whole shared peertube videos, etc.).

How would that look like? When I enter mycustomdomain.org, will the front page be the channel page then? Should my channel be visible on the example.com as well?

**Comment by Jehan** [comment]
> How would that look like? When I enter mycustomdomain.org, will the front page be the channel page then? Should my channel be visible on the example.com as well?

I do believe it makes sense that mycustomdomain.org shows your own channel by default (not the normal default page of example.com). But for more details on customizability, this can be discussed.

Does example.com show your channel as belonging to the "local videos"? I believe that yes. I don't see why you would not want to also appear there (it's more visibility)?

Can visitors reach other people's videos (from the same instance or followed instances) from your domain? You may not want this. Maybe you may also want to have your own list of followed instance instead of example.com's list. Basically you may want to look self-hosted. This could also be options. Typically this is the kind of features which may have a price tag. I.e. an instance server may propose paid accounts for people who want to use their own domain, why not.

How far can you customize appearances for domain hosting? The more the better. Ideally you can even make it look like a completely different peertube instance (except it's the same but would simply load a different CSS when your custom domain is requested, for instance).

Now for me, the real major feature in all this is really about the ability to use one's own domain and also video link stability (you gave a video link to someone 5 years ago, as long as this video has not been removed, it should still be at the same link, even if you changed providers since). This is one of the major points about what "federation" is about IMO: being able to easily change providers thanks to a common protocol. If you have federation but are stuck to a single video hosting provider forever once you chose it, then it's only half federation in my book. Technically you are not stuck, but after you used your account for years and everybody knows your videos are there, in practice you kind of are.

Therefore the most basic implementation of this feature could be: you can attach another domain to a peertube instance, your handle is attached to your domain (not the actual host domain) and when you go there, it shows your channel. Even if there were no other visual customization and if the shown peertube is just the normal example.com, it would already be huge because it means you can move whenever you want. So I would suggest to implement such a basic domain hosting first (then lose ourselves in customizability details later). 🙂

**Comment by test2a** [comment]
Im sorry if I didnt get it correctly but are you suggesting a managed server? Something like https://masto.host/? 

Why dont you talk to those people, maybe bring them to the discussion here? They certainly have an experience with hosting mastodon so Peertube shouldn't be a problem for them

**Comment by Jehan** [comment]
Well maybe Mastodon has the feature, but we are talking about having it in Peertube here. 😉

I have no idea what this service you link does exactly, but it looks to me, they propose to host a full instance. This is **not** what I am talking about. Yes people can already propose a similar service with Peertube. I am proposing a shared service concept, which is an in-between approach. Only one instance but you can use your own domain (not several fully separate instances). From your point of view, you still have your simple account, maybe just a bit more customizability (as discussed above), maybe you have the ability to create other users under your domain (often within a limit for shared services) but not so much more. You are not admin of the instance.

Also this type of features has existed for dozens of years (shared web server with your own domain, shared email provider with your own domain, shared XMPP server with your own domain…), it's nothing extraordinary. I don't think we need really to have input from "people with experience" on shared hosting here. It's like what the internet has done ever since it existed. 😛

**Comment by test2a** [comment]
OK. So you are saying there be a server, websitexyz with abc as admin

You the user wants a slice. You supply your domain and suddenly you are owner of websitesdf which is inreality a subset of websitexyz. You have your own admins or not, you have your own users or not and you pay that abc a monthly fees for the hosting/maintenance/moderatorship.???

**Comment by Jehan** [comment]
No. You are not "owner" of a subset of the hosting website and you don't have any admins. You are a user like anyone else, but you can be by using your own domain, that's all.

I mean, it's just like email providing service. Have you never seen hosting services which allow you to host your emails with your own domain? They don't set up a dedicated email server just for you. You just use the common infrastructure as everyone else but simply have an email with your personal domain name. Probably every email hosting company in the world provides this service. This allows simplicity but in the same time independence (you can leave whenever you want, take all your emails and keep the same address while going to another email provider or start hosting yourself).

P.S.: as for the "monthly fee" part, it was just an example as I'm guessing some people may be interested as hosters. Also as a user, it is also a valuable option. Like paying 10€ a year to have your emails hosted on a shared service is worth not having to care about administration at all while still using your own domain name.
I don't think the "fee" part should be a part of the software, it's up to whatever hosting service to decide how they handle such feature (enable for free, disable for everyone, enable for a fee, whatever…). Was for the sake of example. 🙂
